### DISCUSSION TITLE
How can I get rootdir or absolute paths of the discovered tests?

### DISCUSSION THREAD

**Discussion by misrasaurabh1** [pytest-dev/pytest#12686] [Q&A] [answered]
Hi all,

I am building a tool that runs external users tests to verify code changes. We want to discover the absolute paths of the tests that pytest discovers. Since if you add the rootdir path to the collected  test path from `--co -q`,  the paths become absolute and this is the logic we were using until now.  We want to find rootdir so that we can find the absolute path of the discovered test files like `absolute_test_path = os.path.normpath(os.path.join(pytest_rootdir, test_file))`.

I haven't found a good solution to find rootdir but to just use `--co` run and then parse the rootdir from the printed header. But this can be flaky when users can customize their pytest setup. For some inexplicable reasons I have seen the header be missing.

Is there a better way to directly get rootdir for a test setup before executing all the tests? Better yet, is there a reliable way to get the absolute paths of the tests that are discovered?

Thank you in advance

**Comment by RonnyPfannschmidt** [comment]
Your use case screens use a actual plugin instead of output parsing 

All data is readily available on the python side

**Reply by nicoddemus** [reply, marked as answer]
To complement @RonnyPfannschmidt's answer, use a custom plugin and implement `pytest_collect_modifyitems`: each item will have its full path available then. It is also much more stable/safer than parsing the output, as the hook is part of the public/stable API.

**Reply by misrasaurabh1** [reply]
Thank you for suggesting this answer! We got the prototype working and we will move our code to this method.
